### ISSUE TITLE
Replace CRSF token with SameSite=strict

### ISSUE THREAD

**Issue by silverwind** [go-gitea/gitea#11188]
[`SameSite=strict`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite) effectively prevents Cookie-based CRSF attacks and it also brings the benefit of simplifying our code. From [Wikipedia](https://en.wikipedia.org/wiki/Cross-site_request_forgery#SameSite_cookie_attribute):

> An additional "SameSite" attribute can be included when the server sets a cookie, instructing the browser on whether to attach the cookie to cross-site requests. If this attribute is set to "strict", then the cookie will only be sent on same-origin requests, making CSRF ineffective.

Browser support is [pretty good](https://caniuse.com/#feat=same-site-cookie-attribute) on it. It also means cookies will never be send to other domains like when `STATIC_URL_PREFIX` is set differently, but as far as I'm aware, cookies are not needed for static assets.

Related: https://github.com/go-gitea/gitea/issues/5583

**Comment by silverwind** [comment]
There is one concerning bit in the [RFC draft](https://tools.ietf.org/html/draft-west-first-party-cookies-07) which distinguishes a "registrable domain" as opposed to origin:

> A request is "same-site" if its target's URI's origin's registrable domain is an exact match for the request's initiator's "site for cookies", and "cross-site" otherwise.

Unless I'm missing something, CRSF from `sub1.example.com` to `sub2.example.com` would still possible with that, thought I guess the risk is much lower in such a scenario.

**Comment by lafriks** [comment]
There is still problem if Gitea is hosted in subdir with other apps

**Comment by silverwind** [comment]
@lafriks indeed, but I'd say that's a rare scenario and most of today's web's security relies on origins so they would be vulnerable to other attack vectors as well.

**Comment by guillep2k** [comment]
I'd rather not rely on draft recommendations for security enforcement. Especially since older browsers will just not implement it.

**Comment by silverwind** [comment]
I don't know why it's still draft but it's implemented since 4 years so it should be rather stable. Browsers are in process of enabling the lax setting for all sites or already have done so (Chrome did).

**Comment by CirnoT** [comment]
There is a very limiting factor to `Strict` cookies - if a link to Gitea issue or PR is provided on external site, clicking it will be seen as cross-site request and user will be logged out upon following it.

**Comment by mohe2015** [comment]
According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie subdomains don't get the cookies if no domain is set.

I think explicitly setting SameSite=Lax should be enough. According to https://web.dev/samesite-cookies-explained/: "The default behaviour applied by Chrome is slightly more permissive than an explicit SameSite=Lax as it will allow certain cookies to be sent on top-level POST requests"

**Comment by CirnoT** [comment]
`Lax` is certainly a good idea to implement alone, but not as direct replacement for proper CSRF tokens.

**Comment by mohe2015** [comment]
According to https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html: "Do not use GET requests for state changing operations."

So I think these occurrences should be changed then.

Non state-changing should be fine AFAIK.

**Comment by silverwind** [comment]
Pretty sure we have some non-semantic GETs which may block us to some degree here. One such example is https://github.com/go-gitea/gitea/issues/11188.

**Comment by DuckDuckWhale** [comment]
#12403 and its original issue #10725 seems to be fixed. Any updates?

**Comment by silverwind** [comment]
Some work had been done to get SameSite support into macaron but I'm not sure where it stands currently as there's also the migration to chi ongoing.

**Comment by zeripath** [comment]
#14900 will make all our cookies SameSite strict by default

**Comment by zeripath** [comment]
I guess the next step would be then to add configuration to drop the csrf

**Comment by CirnoT** [comment]
CSRF tokens should still be observed along with cookie configuration. Removal of it would be a very premature action considering browsers are still not yet fully enforcing new cookie configuration in various edge scenarios as to not break the hell out of everything and we have to deal with multiple LTS releases such as Firefox ESR being used.

ref. https://medium.com/@renwa/bypass-samesite-cookies-default-to-lax-and-get-csrf-343ba09b9f2b

**Comment by mohe2015** [comment]
What about SameSite strict or a combination (lax for GET and both for POST as far as I can remember)?

**Comment by silverwind** [comment]
CRSF removal would be nice to have to eliminate a few bugs we have around it. I'm not sure how those bugs exactly but there are ways CRSF will be rejected by legimitate clients. I guess one way to reproduce may be open form, restart server, send form.

**Comment by jub0bs** [comment]
I'm not part of the Gitea community, but I urge you not to rely solely on `SameSite` to the detriment of anti-CSRF tokens. The `SameSite` cookie attribute is [powerless against cross-origin, same-site attacks](https://jub0bs.com/posts/2021-01-29-great-samesite-confusion/), and many orgs do not scrutinize the security level of their subdomains.

**Comment by mohe2015** [comment]
> I'm not part of the Gitea community, but I urge you not to rely solely on `SameSite` to the detriment of anti-CSRF tokens. The `SameSite` cookie attribute is [powerless against cross-origin, same-site attacks](https://jub0bs.com/posts/2021-01-29-great-samesite-confusion/), and many orgs do not scrutinize the security level of their subdomains.

Great so now we have an RFC with a security feature that is not a security feature. Why is there no `SameOrigin` cookie-option...

**Comment by wxiaoguang** [comment]
FYI: in many cases, "SameSite=strict" can't be used, it will lead to bugs like "Cookie SAME_SITE = strict doens't work with external SSO #27033"
